src/transformers/models/seamless_m4t/modeling_seamless_m4t.py

```diff
@@ -0,0 +1,3944 @@
+        self,
+        input_ids: Optional[torch.Tensor] = None,
+        input_features: Optional[torch.Tensor] = None,
+        return_intermediate_token_ids: Optional[bool] = None,
+        **kwargs,
+    ) -> Union[torch.Tensor, SeamlessM4TGenerationOutput]:
+        vocoder_tgt_lang_id = kwargs.pop("vocoder_tgt_lang_id", None)
+
+        kwargs_text = {"decoder_input_ids": kwargs.pop("decoder_input_ids", None)}
+        kwargs_speech = {}
+        for key, value in kwargs.items():
+            if key.startswith("text_"):
+                key = key[len("text_") :]
+                kwargs_text[key] = value
+            elif key.startswith("speech_"):
+                key = key[len("speech_") :]
+                kwargs_speech[key] = value
+            else:
+                # If the key is already in a specific config, then it's been set with a
+                # submodules specific value and we don't override
+                if key not in kwargs_text:
+                    kwargs_text[key] = value
+                if key not in kwargs_speech:
+                    kwargs_speech[key] = value
+
+        if input_ids is None and input_features is None and kwargs.get("inputs_embeds", None) is None:
+            raise ValueError(
+                "`input_ids`,`input_features` and `inputs_embeds` are all empty. Make sure at least one of them is not."
+            )
+
+        kwargs_text["output_hidden_states"] = True
+        kwargs_text["return_dict_in_generate"] = True
+        kwargs_text["output_scores"] = True
+
+        if input_features is not None:
+            self.set_modality("speech")
+            if input_ids is not None:
+                logger.warning(
+                    "`input_features` and `input_ids` are both non empty. `input_features` will be used in priority through the speech encoder."
+                    "Make sure `input_features=None` if you want to use the text encoder."
+                )
+            text_generation_output = super().generate(input_features=input_features, **kwargs_text)
+            batch_size = len(input_features)
+        else:
+            self.set_modality("text")
+            text_generation_output = super().generate(input_ids=input_ids, input_features=None, **kwargs_text)
+            batch_size = len(input_ids)
+
+        num_return_sequences = len(text_generation_output.sequences) // batch_size
+        sequences = text_generation_output.sequences
+
+        attention_mask = kwargs_speech.get("attention_mask", kwargs_text.get("attention_mask", None))
+
+        if self.current_modality == "speech":
+            # get last_hidden_state from encoder
+            encoder_hidden_states = self.speech_encoder(input_features=input_features, attention_mask=attention_mask)[
+                0
+            ]
+
+            # input modality = speech so new attention mask for the decoder
+            if attention_mask is not None:
+                attention_mask = _compute_new_attention_mask(
```

Comment:
`_compute_new_attention_mask` could become a method like it is in Wav2Vec2 if we register `self.kernel_size = config.adaptor_kernel_size` in the init?